scrapers/utils.py

```diff
@@ -134,3 +142,89 @@ def validate_and_get_download_link(base_url, download_link, check_content_pdf=Tr
             raise TypeError("File extension does not match content type.")
 
     return absolute_url
+
+
+def calculate_checksum(filepaths):
+    md5 = hashlib.md5()
+    for file in filepaths:
+        try:
+            md5.update(Path(file).read_bytes())
+        except IsADirectoryError:
+            pass
+    return md5.hexdigest()
+
+
+def zip_pdfs(zipped_directory, scraper_filepath, filepaths, link_id):
+    zip_name = get_standardized_filename(scraper_filepath, link_id, ".zip")
+    zip_path = get_file_path(zipped_directory, zip_name)
+
+    with zipfile.ZipFile(zip_path, "w") as zipf:
+        for file in filepaths:
+            filename = os.path.basename(file)
+            zipf.write(file, arcname=filename)
+
+    return zip_name, zip_path
+
+
+def connect_circulars_to_associated_files(file_names):
+    """
+    Accepts 'file_names' as an argument, which is a list of lists in which each sub-list consists of a file name at index 0 and a file extension at index 1.
+
+    This function identifies files that are part of the same circular and groups them together in a returned dictionary, in which each key is a circular name and its value is a list of all files belonging to that circular.
+
+    Example of valid input:     [ ["file1", "pdf"], ["file2", "pdf"], ["file3", "docx"] ]   OR    [ ("file1", "pdf") ]
+
+    Invalid input:      [ ["file1", 1] ]    OR    [ {"file1", "pdf"}, {"file2", "pdf"} ]    OR  { "file1": "pdf", "file2": "pdf" }
+    """
+    circular_to_links_dict = {}
+    try:
+        # sort file_names list by file name string length
+        file_names.sort(key=lambda x: len(x[0]))
+    except Exception as e:
+        raise TypeError(f"Invalid 'file_names' argument.")
+
+    for i, file_list in enumerate(file_names):
+        try:
+            curr_file, extension = file_list
+            if (
+                not isinstance(curr_file, str)
+                or not isinstance(extension, str)
+                or (not isinstance(file_list, list) and not isinstance(file_list, tuple))
+            ):
+                raise
+        except Exception:
+            raise ValueError(f"The element at index {i} is invalid.")
```

Comment:
That is not the correct use of try/except block. Also, if `file_list` is not in the format that we expect the way we won't be able to unpack the values properly anyway. I would not worry about that.

The moral of the story here is that trying to raise custom exception every time is not the correct approach. Sometimes allowing the operation to raise their own exception will facilitate easier debugging - not always though! So it's a bit of a trial and error and experience to understand what is what.

```suggestion
            curr_file, extension = file_list
            if (
                not isinstance(curr_file, str)
                or not isinstance(extension, str)
            ):
                raise TypeError(f"Incorrect type for file name and extension.")
```

Reply:
This makes sense. I had the feeling that the default error handling might be more beneficial here due to the specificity.